### PR TITLE
pkg/boot: Use a library for xz files

### LINKED ISSUE #2750: image.Load tries using system xz binary
**Is your feature request related to a problem? Please describe.**

I have this code to use kexec to install the latest version of my image on machines that are outdated.

```
func loadImage(image *boot.LinuxImage, verbose bool) error {
	err := image.Load(boot.WithVerbose(verbose))
	if err != nil {
		log.Printf("Failed to load kernel: %v. Retrying with LoadSyscall=true.", err)
		image.LoadSyscall = true
		err = image.Load(boot.WithVerbose(verbose))
	}
	if err != nil {
		return fmt.Errorf("Kexec load of %v failed: %v", image.Label(), err)
	}
	return nil
}
```
The reason I have the Load bit twice is that some of the machines have a kernel that's compiled without the `CONFIG_KEXEC_FILE` flag.

I'm hitting this error on some machines:

```
Failed to load image: Kexec load of Installer failed: parsing bzImage Linux kernel: error decompressing payload: error starting decompressor: exec: "unxz": executable file not found in $PATH
````

Is there a way to bundle the needed libraries into u-root itself instead of relying on the machine having this binary?

**Describe the solution you'd like**
The needed libraries would be bundled into u-root so that it doesn't have to shell out to the system.

### PR DESCRIPTION
**Pull request by navidpaya**
Fixes #2750

At the moment, we shell out for decompressing xz files. This change will change that to using a library instead.

Tested this on my local code.

Before:

```
parsing bzImage Linux kernel: error decompressing payload: failed writing decompressed bytes to writer: xz: data is truncated or corrupt
```

```
found decompressor for signature:  [253 55 122 88 90 0]
```

The code managed to kexec using the same kernel with the code that has as the fix.

### CODE REVIEW

File: pkg/boot/bzimage/bzimage_decompress.go [hunk 1]
```diff
@@ -12,6 +12,7 @@ import (
 	"io"
 	"os/exec"
 
+	"github.com/therootcompany/xz/v/v1/xz"
```
Comment (hugelgupf) [inline pkg/boot/bzimage/bzimage_decompress.go:15]: if we're going to depend on an xz library, let's not depend on a new one. we already have github.com/ulikunitz/xz included below?

File: go.mod [hunk 2]
```diff
@@ -31,6 +31,7 @@ require (
 	github.com/rekby/gpt v0.0.0-20200219180433-a930afbc6edc
 	github.com/safchain/ethtool v0.0.0-20200218184317-f459e2d13664
 	github.com/spf13/pflag v1.0.5
+	github.com/therootcompany/xz v1.0.1
```
Comment (10000TB) [inline go.mod:34]: any reason for using this forked version, than the original https://github.com/xi2/xz ?

File: pkg/boot/bzimage/bzimage_decompress.go [hunk 3]
```diff
@@ -127,3 +128,31 @@ func unzstd(w io.Writer, r io.Reader) error {
...
+	if err != nil {
+		return fmt.Errorf("error creating unxz reader: %v", err)
+	}
+
+	if _, err := io.Copy(w, unxzReader); err != nil {
+		return fmt.Errorf("failed writing decompressed bytes to writer: %v", err)
+	}
+	return nil
+}
+
```
Comment (10000TB) [inline pkg/boot/bzimage/bzimage_decompress.go:145]: looks like duplicate.

**Review by rminnich — CHANGES_REQUESTED** [review]
hang on. 

we shelled out because we could not handle files compressed with BCJ, for example. I just want to be sure that we're not losing something here.

**Comment by rminnich** [comment]
the other option is to use the cpu command, with a container for the target architecture, such that you can have the unxz you need. I do that all the time. 

Not sure I trust that org just yet, This site can’t provide a secure connectionrootprojects.org sent an invalid response.
ERR_SSL_PROTOCOL_ERROR

**Comment by rminnich** [comment]
another option is to have a build tag that enables this library, with default off?

File: go.mod (hunk 2 above)
Comment (navidpaya) [reply go.mod:34]: Licensing. https://github.com/xi2/xz/blob/master/LICENSE doesn't mention an explicit license but https://github.com/therootcompany/xz/blob/master/LICENSE uses CC0 1.0 which is good enough for company's repository system to be happy with pulling it in.

**Comment by navidpaya** [comment]
> hang on.
> 
> we shelled out because we could not handle files compressed with BCJ, for example. I just want to be sure that we're not losing something here.

I can't comment to the original issue but can say that testing the image we use, the modified code managed to kexec so seems to be working.

**Comment by navidpaya** [comment]
> the other option is to use the cpu command, with a container for the target architecture, such that you can have the unxz you need. I do that all the time.
> 
> Not sure I trust that org just yet, This site can’t provide a secure connectionrootprojects.org sent an invalid response. ERR_SSL_PROTOCOL_ERROR

Can you provide a link for that? I'm not familiar with that method but could look into it.

The main issue here is that we use u-root for reinstalling machines that are in a weird state so can't reliably use the machine and have to rely on our u-root binary to take care of everything (download the artifacts and kexec into the new kernel).

**Comment by navidpaya** [comment]
> another option is to have a build tag that enables this library, with default off?

I could look into that if you folks are comfortable with that. If there's an example of that, that would be superb.

**Comment by rminnich** [comment]
https://rootprojects.org/ gets me an error.

"This site can’t provide a secure connectionrootprojects.org sent an invalid response.
ERR_SSL_PROTOCOL_ERROR"

So I worry about about the longevity of that org. But the go proxy should keep us safe there I guess.

That said, I am thinking I was wrong, and the code would be nice. It's a tad worrisome because this is load-bearing at google and other companies, and we don't want to break anybody's data center :-)

But I hate that we have to shell out to things like unxz. 

so, how do people feel about just going with this change? I'm leaning to "let's go". Anybody? If that fails, I guess we can go the build tag route.

navidpaya, thanks for your patience here, we've all been burned by kexec one time or another so we're careful.

I can think of another option, where we try the package uncompressor, if it fails, we shell out to unxz. That's pretty easy to put in the code. Comments?

**Comment by navidpaya** [comment]
> https://rootprojects.org/ gets me an error.
> 
> "This site can’t provide a secure connectionrootprojects.org sent an invalid response. ERR_SSL_PROTOCOL_ERROR"
> 
> So I worry about about the longevity of that org. But the go proxy should keep us safe there I guess.
> 
> That said, I am thinking I was wrong, and the code would be nice. It's a tad worrisome because this is load-bearing at google and other companies, and we don't want to break anybody's data center :-)
> 
> But I hate that we have to shell out to things like unxz.
> 
> so, how do people feel about just going with this change? I'm leaning to "let's go". Anybody? If that fails, I guess we can go the build tag route.
> 
> navidpaya, thanks for your patience here, we've all been burned by kexec one time or another so we're careful.
> 
> I can think of another option, where we try the package uncompressor, if it fails, we shell out to unxz. That's pretty easy to put in the code. Comments?

No worries, Ron. Fellow Googler here. I definitely wouldn't want to break u-root either :)

I'm going to make the change to try the library and fall back to shelling out as well as the other changes needed to make the tests pass.

File: pkg/boot/bzimage/bzimage_decompress.go [hunk 4]
```diff
@@ -127,3 +128,18 @@ func unzstd(w io.Writer, r io.Reader) error {
...
 	return nil
 }
+
+// unxz reads compressed bytes from the io.Reader and writes the uncompressed bytes to the
+// writer. unxz satisfies the decompressor interface.
+func unxz(w io.Writer, r io.Reader) error {
+	unxzReader, err := xz.NewReader(r, 0)
+	if err != nil {
+		// falling back to shelling out to 'unxz' 
+		unxzReader, nil = execer("unxz")
```
Comment (navidpaya) [inline pkg/boot/bzimage/bzimage_decompress.go:138]: Looks like I totally misunderstood how to implement that. Anyone knows where I should implement falling back to shell? I'm guessing in bzimage.go but all those magic bits are inline.

File: pkg/boot/bzimage/bzimage_decompress.go (hunk 3 above)
Comment (navidpaya) [reply pkg/boot/bzimage/bzimage_decompress.go]: Thanks, removed

File: pkg/boot/bzimage/bzimage_decompress.go (hunk 4 above)
Comment (navidpaya) [reply pkg/boot/bzimage/bzimage_decompress.go]: I implemented the fixes (thanks for the help @rminnich). We now have fallback to the xz binary. The tests are passing too.

**Comment by navidpaya** [comment]
I'm going to close this in favor of  #3052. I noticed my git comments are quite messy and couldn't find a way to fix that.

**Pull request closed without merging**